homework/index.js

```diff
@@ -29,23 +31,115 @@
+      text: repo.name,
+      href: repo.html_url,
+      target: '_blank',
+    });
+    renderRepoTable(table, 'Description:', repo.description);
+    renderRepoTable(table, 'Forks:', repo.forks);
+    const date = new Date(repo.updated_at).toLocaleString('en-US');
+    renderRepoTable(table, 'Updated:', date);
+  }
+
+  function renderContributors(url, div) {
+    fetchJSON(url)
+      .then(repos => {
+        repos.forEach(repo => {
+          const divContributor = createAndAppend('div', div, {
+            class: 'contributor-details',
+          });
+          const linkContributor = createAndAppend('a', divContributor, {
+            href: repo.html_url,
+            target: '_blank',
+          });
+          createAndAppend('img', linkContributor, {
+            src: repo.avatar_url,
+            class: 'avatar',
+          });
+          createAndAppend('h4', linkContributor, {
+            text: repo.login,
+          });
+          createAndAppend('p', linkContributor, {
+            text: repo.contributions,
+            class: 'contribution-count',
+          });
+        });
+      })
+      .catch(err => {
+        createAndAppend('div', div, {
+          text: err.message,
+          class: 'alert-error',
+        });
+      });
   }
 
   function main(url) {
-    fetchJSON(url, (err, repos) => {
-      const root = document.getElementById('root');
-      if (err) {
-        createAndAppend('div', root, {
+    fetchJSON(url)
+      .then(repos => {
+        const root = document.getElementById('root');
+        const header = createAndAppend('header', root, {
+          class: 'header-class',
+        });
+        createAndAppend('h3', header, {
+          text: 'HYF Repository',
+          class: 'title',
+        });
+        const select = createAndAppend('select', header, {
+          class: 'select-class',
+        });
+        const mainContainer = createAndAppend('section', root, {
+          class: 'main-container',
+        });
+        const reposContainer = createAndAppend('section', mainContainer, {
+          class: 'repos-container',
+        });
+        const contributorContainer = createAndAppend('section', mainContainer, {
+          class: 'contributor-container',
+        });
+        repos
+          .sort((a, b) => {
+            return a.name.localeCompare(b.name);
+          })
+          .forEach((elem, index) => {
+            createAndAppend('option', select, {
+              value: index,
+              text: repos[index].name,
+              class: 'select-class',
+            });
+          });
+        function renderContent(elem) {
+          reposContainer.innerHTML = '';
+          contributorContainer.innerHTML = '';
+          const repo = repos[elem];
+          renderRepoDetails(repo, reposContainer);
+          contributorContainer.innerHTML = 'Contributors';
+          renderContributors(repo.contributors_url, contributorContainer);
+        }
```

Comment:
You have correctly understood that you can create this function to make your code DRY. However, you avoid embedded named functions like this, because it is not easy to determine to which variables from the outer function they need access. Furthermore, it increases the size of the outer function, which makes it less readable.

If you move this function out of the `main` function, you will immediately see in VSCode various line having red underlines, indicating issues. Notably, this is the case with the variables `reposContainer`, `contributorContainer`, and `repos`.

```js
function renderContent(elem) {
  reposContainer.innerHTML = '';
  contributorContainer.innerHTML = '';
  const repo = repos[elem];
  renderRepoDetails(repo, reposContainer);
  contributorContainer.innerHTML = 'Contributors';
  renderContributors(repo.contributors_url, contributorContainer);
}

function main(url) {
  ...
}
```

Therefore, you need to add the missing variables as parameters to the function definition:

```js
function renderContent(elem, repos, reposContainer, contributorContainer) {
  ...
}
```

But if you look closer, you can see that you are using `elem` and `repos` inside the function only to get the repo that is to be rendered:

```js
const repo = repos[elem];
```

Therefore, it is better to let the caller of the function pass this repo as a parameter so that you can simplify the function definition to:

```js
function renderContent(repo, reposContainer, contributorContainer) {
  reposContainer.innerHTML = '';
  contributorContainer.innerHTML = '';
  renderRepoDetails(repo, reposContainer);
  contributorContainer.innerHTML = 'Contributors';
  renderContributors(repo.contributors_url, contributorContainer);
}
```

Then, in your main function, you can call this function with:

```js
renderContent(repos[0], reposContainer, contributorContainer)
```

Reply:
I removed the function outside the main function, and rendered with :

renderContent(repos[0], reposContainer, contributorContainer);
renderContent(repos[select.value], reposContainer, contributorContainer)